README.md
# react-native-datepicker [![Build Status](https://travis-ci.org/xgfe/react-native-datepicker.svg?branch=master)](https://travis-ci.org/xgfe/react-native-datepicker) [![Coverage Status](https://coveralls.io/repos/github/xgfe/react-native-datepicker/badge.svg?branch=master)](https://coveralls.io/github/xgfe/react-native-datepicker?branch=master) [![Monthly download](https://img.shields.io/npm/dm/react-native-datepicker.svg)](https://img.shields.io/npm/dm/react-native-datepicker.svg) [![Total downloads](https://img.shields.io/npm/dt/react-native-datepicker.svg)](https://img.shields.io/npm/dt/react-native-datepicker.svg)
React Native DatePicker component for both Android and iOS, using DatePickerAndroid, TimePickerAndroid and DatePickerIOS

## Install

```bash
npm install react-native-datepicker --save
npm install @react-native-community/datetimepicker --save
```

Or using [react-native-ui-xg](https://github.com/xgfe/react-native-ui-xg), our react-naitve ui kit.

```bash
npm install react-native-ui-xg --save
```

## iOS 13 Dark Mode

Currently there is an issue with dark mode and date picker on iOS 13 and app built on Xcode 11 or higher. To workaround this issue at the moment you need to add you your iOS project plist file o opt-out from dark mode. Future versons of date picker or React Native should resolve it correctly.

```xml
<key>UIUserInterfaceStyle</key>
<string>Light</string>
```

## Example
Check [index.android.js](https://github.com/xgfe/react-native-datepicker/blob/master/index.android.js) in the Example.

![android](http://xgfe.github.io/react-native-datepicker/img/react-native-datepicker-android.gif)
![ios](http://xgfe.github.io/react-native-datepicker/img/react-native-datepicker-ios.gif)

## Usage

```javascript

import React, { Component } from 'react'
import DatePicker from 'react-native-datepicker'

export default class MyDatePicker extends Component {
  constructor(props){
    super(props)
    this.state = {date:"2016-05-15"}
  }

  render(){
    return (
      <DatePicker
        style={{width: 200}}
        date={this.state.date}
        mode="date"
        placeholder="select date"
        format="YYYY-MM-DD"
        minDate="2016-05-01"
        maxDate="2016-06-01"
        confirmBtnText="Confirm"
        cancelBtnText="Cancel"
        customStyles={{
          dateIcon: {
            position: 'absolute',
            left: 0,
            top: 4,
            marginLeft: 0
          },
          dateInput: {
            marginLeft: 36
          }
          // ... You can check the source to find the other keys.
        }}
        onDateChange={(date) => {this.setState({date: date})}} // Called when the user confirm the picked date or time in the UI 
      />
    )
  }
}
```

You can check [index.js](https://github.com/xgfe/react-native-datepicker/blob/master/index.android.js) in the Example for detail.

## Properties

| Prop  | Default  | Type | Description |
| :------------ |:---------------:| :---------------:| :-----|
| style | - | `object` | Specify the style of the DatePicker, eg. width, height...  |
| date | - | <code>string &#124; date &#124; Moment instance</code> | Specify the display date of DatePicker. `string` type value must match the specified format |
| mode | 'date' | `enum` | The `enum` of `date`, `datetime` and `time` |
| androidMode | 'default' | `enum` | The `enum` of `default`, `calendar` and `spinner` (only Android) |
| format | 'YYYY-MM-DD' | `string` | Specify the display format of the date, which using [moment.js](http://momentjs.com/). The default value change according to the mode. |
| confirmBtnText | '确定' | `string` | Specify the text of confirm btn in ios. |
| cancelBtnText | '取消' | `string` | Specify the text of cancel btn in ios. |
| iconSource | - | <code>{uri: string} &#124; number</code> | Specify the icon. Same as the `source` of Image, always using `require()` |
| minDate | - | <code>string &#124; date</code> | Restricts the range of possible date values. |
| maxDate | - | <code>string &#124; date</code> | Restricts the range of possible date values. |
| duration | 300 | `number` | Specify the animation duration of datepicker.|
| customStyles | - | `object` | The hook of customize datepicker style, same as the native style. `dateTouchBody`, `dateInput`...|
| showIcon | true | `boolean` | Controller whether or not show the icon |
| hideText | false | `boolean` | Controller whether or not show the `dateText` |
| iconComponent | - | `element` | Set the custom icon |
| disabled | false | `boolean` | Controller whether or not disable the picker |
| is24Hour | - | `boolean` | Set the TimePicker is24Hour flag. The default value depend on `format`. Only work in Android |
| allowFontScaling | true | `boolean` | Set to false to disable font scaling for every text component |
| placeholder | '' | `string` | The placeholder show when this.props.date is falsy |
| onDateChange | - | `function` | This is called when the user confirm the picked date or time in the UI. The first and only argument is a date or time string representing the new date and time formatted by [moment.js](http://momentjs.com/) with the given format property. |
| onOpenModal | - | `function` | This is called when the DatePicker Modal open. |
| onCloseModal | - | `function` | This is called when the DatePicker Modal close |
| onPressMask | - | `function` | This is called when clicking the ios modal mask |
| modalOnResponderTerminationRequest | - | `function` | Set the callback for React Native's [Gesture Responder System](https://facebook.github.io/react-native/docs/gesture-responder-system.html#responder-lifecycle)'s call to `onResponderTerminationRequest`. By default this will reject a termination request, but can be overidden in case the View under the Modal is implementing custom gesture responders, and you wish for those to be overidden in certain cases.  |
| TouchableComponent | `TouchableHighlight` | `Component` | Replace the `TouchableHighlight` with a custom `Component`. For example : `TouchableOpacity` |
| getDateStr | - | Function | A function to override how to format the date into a `String` for display, receives a `Date` instance

### Property `customStyles` available keys

* appearance: `dateInput`, `disabled`, `dateTouchBody`, `dateIcon`, `placeholderText`, `dateText`
* ios select panel: `datePickerCon`, `datePicker`, `btnConfirm`, `btnTextConfirm`, `btnCancel`, `btnTextCancel`


## Instance Methods

| Method  | Params  | Description |
| :------------ |:---------------:| :---------------:|
| onPressDate | - | Manually open the date picker panel |
| onPressCancel | - | Manually close the date picker panel like, similarly pressing cancel btn |


CHANGELOG.md
## 1.7.3

Use @react-native-community/datetimepicker insead of deprecated ones.
You need to install @react-native-community/datetimepicker (read install notes)

Add info about iOS 13 Dark Mode issue

<a name="1.7.2"></a>
## [1.7.2](https://github.com/xgfe/react-native-datepicker/compare/v1.7.0...v1.7.2) (2018-04-22)


### Bug Fixes

* **datepicker:** fix androidMode enum not support 'clock' bug ([dc29a09](https://github.com/xgfe/react-native-datepicker/commit/dc29a09))



<a name="1.7.0"></a>
# [1.7.0](https://github.com/xgfe/react-native-datepicker/compare/v1.6.0...v1.7.0) (2018-04-08)


### Bug Fixes

* **datepicker:** fix warns when pass moment instance to date prop ([7450bba](https://github.com/xgfe/react-native-datepicker/commit/7450bba)), closes [#166](https://github.com/xgfe/react-native-datepicker/issues/166)


### Features

* **datepicker:** add `getDateStr` prop to customize display outside moment format ([905dfbd](https://github.com/xgfe/react-native-datepicker/commit/905dfbd))
* **datepicker:** pass `locale` to native control if given ([60e3887](https://github.com/xgfe/react-native-datepicker/commit/60e3887))
* **datepicker:** merge some PR for bugs ([5377499](https://github.com/xgfe/react-native-datepicker/commit/5377499))



<a name="1.6.0"></a>
# [1.6.0](https://github.com/xgfe/react-native-datepicker/compare/v1.5.1...v1.6.0) (2017-06-12)

### Features

* **datepicker:** adding testIds to elements ([#138](https://github.com/xgfe/react-native-datepicker/issues/138)
* **datepicker:** add `hideText` property to hide the Date Text ([#127](https://github.com/xgfe/react-native-datepicker/issues/127)
* **datepicker:** add `TouchableComponent` property to support custom touchable components ([#128](https://github.com/xgfe/react-native-datepicker/issues/128)


<a name="1.5.1"></a>
## [1.5.1](https://github.com/xgfe/react-native-datepicker/compare/v1.5.0...v1.5.1) (2017-05-16)


### Features

* **datepicker:** change datepicker switch gap timeout to 200ms ([5b6e78f](https://github.com/xgfe/react-native-datepicker/commit/5b6e78f))

### Bug Fixes

* **datepicker:** Fix onDateChange for ios ([#123](https://github.com/xgfe/react-native-datepicker/issues/123), [e4e2193](https://github.com/xgfe/react-native-datepicker/commit/e4e2193))


<a name="1.5.0"></a>
# [1.5.0](https://github.com/xgfe/react-native-datepicker/compare/v1.4.7...v1.5.0) (2017-05-15)


### Features

* **datepicker:** Add onPressMask prop for ios ([#120](https://github.com/xgfe/react-native-datepicker/issues/120), [09ecdf8](https://github.com/xgfe/react-native-datepicker/commit/09ecdf8))
* **datepicker:** Ability to pass icon component with `iconComponent` property ([#124](https://github.com/xgfe/react-native-datepicker/issues/124), [35150ed](https://github.com/xgfe/react-native-datepicker/commit/35150ed))

### Bug Fixes

* **datepicker:** Fix onCloseModal for android ([#122](https://github.com/xgfe/react-native-datepicker/issues/122), [ba55c06](https://github.com/xgfe/react-native-datepicker/commit/ba55c06))


<a name="1.4.7"></a>
## [1.4.7](https://github.com/xgfe/react-native-datepicker/compare/v1.4.6...v1.4.7) (2017-04-13)


### Bug Fixes

* **datepicker:** fix `androidMode` prop pass bug ([#111](https://github.com/xgfe/react-native-datepicker/issues/111), [095f756](https://github.com/xgfe/react-native-datepicker/commit/095f756))


<a name="1.4.6"></a>
## [1.4.6](https://github.com/xgfe/react-native-datepicker/compare/v1.4.5...v1.4.6) (2017-04-05)


### Features

* **datepicker:** add `androidMode` property for DatePickerAndroid & TimePickerAndroid ([be30731](https://github.com/xgfe/react-native-datepicker/commit/be30731))



<a name="1.4.5"></a>
## [1.4.5](https://github.com/xgfe/react-native-datepicker/compare/v1.4.4...v1.4.5) (2017-03-21)


### Bug Fixes

* **datepicker:** fix `date` prop change not invoke state change bug ([#93](https://github.com/xgfe/react-native-datepicker/issues/93), [3c8099b](https://github.com/xgfe/react-native-datepicker/commit/3c8099b))



<a name="1.4.4"></a>
## [1.4.4](https://github.com/xgfe/react-native-datepicker/compare/v1.4.3...v1.4.4) (2017-02-06)


### Features

* **datepicker:** Added two new properties: onOpenModal and onCloseModal ([#77](https://github.com/xgfe/react-native-datepicker/pull/77), [846ae35](https://github.com/xgfe/react-native-datepicker/commit/846ae35))
* **Datepicker:** Add animation on datepicker dismiss ([#76](https://github.com/xgfe/react-native-datepicker/pull/76), [fdfcc21](https://github.com/xgfe/react-native-datepicker/commit/fdfcc21))

### Bug Fixes

* **datepicker:** fix datepicker can't  open in landscape mode bug ([#28](https://github.com/xgfe/react-native-datepicker/issues/28), [3a9df99](https://github.com/xgfe/react-native-datepicker/commit/3a9df99))



<a name="1.4.3"></a>
## [1.4.3](https://github.com/xgfe/react-native-datepicker/compare/v1.4.1...v1.4.3) (2016-12-13)


### Bug Fixes

* **datepicker:** fix `disabled` props change not update bug [#66](https://github.com/xgfe/react-native-datepicker/issues/66) ([7579dc9](https://github.com/xgfe/react-native-datepicker/commit/7579dc9)), closes [#66](https://github.com/xgfe/react-native-datepicker/issues/66)
* **datepicker:** fix is24Hour flag can't be specified bug ([c320e56](https://github.com/xgfe/react-native-datepicker/commit/c320e56)), closes [#69](https://github.com/xgfe/react-native-datepicker/issues/69)


### Reverts

* **datepicker:** revert responder prevert func temporary [#47](https://github.com/xgfe/react-native-datepicker/issues/47) [#50](https://github.com/xgfe/react-native-datepicker/issues/50) ([5b9f60c](https://github.com/xgfe/react-native-datepicker/commit/5b9f60c))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/xgfe/react-native-datepicker/compare/v1.4.0...v1.4.1) (2016-11-03)


### Bug Fixes


datepicker.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  Text,
  Image,
  Modal,
  TouchableHighlight,
  Platform,
  Animated,
  Keyboard,
} from 'react-native';
import Style from './style';
import Moment from 'moment';
import RNDateTimePicker from '@react-native-community/datetimepicker';

const FORMATS = {
  'date': 'YYYY-MM-DD',
  'datetime': 'YYYY-MM-DD HH:mm',
  'time': 'HH:mm',
};

const SUPPORTED_ORIENTATIONS = ['portrait', 'portrait-upside-down', 'landscape', 'landscape-left', 'landscape-right'];

class DatePicker extends Component {
  constructor(props) {
    super(props);

    this.state = {
      date: this.getDate(),
      modalVisible: false,
      androidDatePickerVisible: false,
      androidTimePickerVisible: false,
      animatedPosition: new Animated.Value(0),
      allowPointerEvents: true,
    };

    this.getDate = this.getDate.bind(this);
    this.getDateStr = this.getDateStr.bind(this);
    this.datePicked = this.datePicked.bind(this);
    this.onPressDate = this.onPressDate.bind(this);
    this.onPressCancel = this.onPressCancel.bind(this);
    this.onPressConfirm = this.onPressConfirm.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.onPressMask = this.onPressMask.bind(this);
    this.onDatePicked = this.onDatePicked.bind(this);
    this.onTimePicked = this.onTimePicked.bind(this);
    this.onDatetimePicked = this.onDatetimePicked.bind(this);
    this.onDatetimeTimePicked = this.onDatetimeTimePicked.bind(this);
    this.setModalVisible = this.setModalVisible.bind(this);
    this.setAndroidDatePickerVisible = this.setAndroidDatePickerVisible.bind(this);
    this.setAndroidTimePickerVisible = this.setAndroidTimePickerVisible.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.date !== this.props.date) {
      this.setState({date: this.getDate(this.props.date)});
    }
  }

  setModalVisible(visible) {
    const { height, duration } = this.props;

    // slide animation
    if (visible) {
      this.setState({modalVisible: visible});
      return Animated.timing(
        this.state.animatedPosition,
        {
          toValue: -height,
          duration: duration,
          useNativeDriver: true,
        },
      ).start();
    } else {
      return Animated.timing(
        this.state.animatedPosition,
        {
          toValue: 0,
          duration: duration,
          useNativeDriver: true,
        },
      ).start(() => {
        this.setState({modalVisible: visible});
      });
    }
  }

  setAndroidDatePickerVisible(visible) {
    this.setState( { androidDatePickerVisible: visible });
  }

  setAndroidTimePickerVisible(visible) {
    this.setState( { androidTimePickerVisible: visible });
  }

  onStartShouldSetResponder(e) {
    return true;
  }

  onMoveShouldSetResponder(e) {
    return true;
  }

  onPressMask() {
    if (typeof this.props.onPressMask === 'function') {
      this.props.onPressMask();
    } else {
      this.onPressCancel();
    }
  }

  onPressCancel() {
    this.setModalVisible(false);

    if (typeof this.props.onCloseModal === 'function') {
      this.props.onCloseModal();
    }
  }

  onPressConfirm() {
    this.datePicked();
    this.setModalVisible(false);

    if (typeof this.props.onCloseModal === 'function') {
      this.props.onCloseModal();
    }
  }

  getDate(date = this.props.date) {
    const { mode, minDate, maxDate, format = FORMATS[mode] } = this.props;

    // date
    if (!date) {
      let now = new Date();
      if (minDate) {
        let _minDate = this.getDate(minDate);

        if (now < _minDate) {
          return _minDate;
        }
      }

      if (maxDate) {
        let _maxDate = this.getDate(maxDate);

        if (now > _maxDate) {
          return _maxDate;
        }
      }

      return now;
    }

    if (date instanceof Date) {
      return date;
    }

    return Moment(date, format).toDate();
  }

  getDateStr(date = this.props.date) {
    const { mode, format = FORMATS[mode] } = this.props;

    const dateInstance = date instanceof Date
      ? date
      : this.getDate(date);

    if (typeof this.props.getDateStr === 'function') {
      return this.props.getDateStr(dateInstance);
    }

    return Moment(dateInstance).format(format);
  }

  datePicked() {
    if (typeof this.props.onDateChange === 'function') {
      this.props.onDateChange(this.getDateStr(this.state.date), this.state.date);
    }
  }

  getTitleElement() {
    const {date, placeholder, customStyles, allowFontScaling} = this.props;

    if (!date && placeholder) {
      return (
        <Text allowFontScaling={allowFontScaling} style={[Style.placeholderText, customStyles.placeholderText]}>
          {placeholder}
        </Text>
      );
    }
    return (
      <Text allowFontScaling={allowFontScaling} style={[Style.dateText, customStyles.dateText]}>
        {this.getDateStr()}
      </Text>
    );
  }

  onDateChange(event, date) {
    this.setState({
      allowPointerEvents: false,
      date: Moment(date).toDate(),
    });
    const timeoutId = setTimeout(() => {
      this.setState({
        allowPointerEvents: true,
      });
      clearTimeout(timeoutId);
    }, 200);
  }

  onDatePicked(event, date) {
    this.setAndroidDatePickerVisible(false);
    if (date === undefined) {
      this.onPressCancel();
    } else {
      this.setState({
        date: Moment(date).toDate(),
      });
      this.datePicked();
    }
  }

  onTimePicked(event, date) {
    this.setAndroidTimePickerVisible(false);
    if (date === undefined) {
      this.onPressCancel();
    } else {
      this.setState({
        date: Moment(date).toDate(),
      });
      this.datePicked();
    }
  }

  onDatetimePicked(event, date) {
    const { mode, androidMode, format = FORMATS[mode], is24Hour = !format.match(/h|a/)} = this.props;

    const newDate = Moment(date);
    const oldDate = Moment(this.state.date);
    newDate.hours(oldDate.hours());
    newDate.minutes(oldDate.minutes());
    newDate.seconds(oldDate.seconds());
    newDate.milliseconds(oldDate.milliseconds());

    this.setAndroidDatePickerVisible(false);
    if (date === undefined) {
      this.onPressCancel();
    } else {
      this.setState({
        date: newDate.toDate(),
      });
      this.setAndroidTimePickerVisible(true);
    }
  }

  onDatetimeTimePicked(event, date) {

    const newDate = Moment(date);
    const dateToSave = Moment(this.state.date);
    dateToSave.hours(newDate.hours());
    dateToSave.minutes(newDate.minutes());
    dateToSave.seconds(newDate.seconds());
    dateToSave.milliseconds(newDate.milliseconds());

    this.setAndroidTimePickerVisible(false);
    if (date === undefined) {
      this.onPressCancel();
    } else {
      this.setState({
        date: dateToSave.toDate(),
      });
      this.datePicked();
    }
  }

  onPressDate() {
    if (this.props.disabled) {
      return true;
    }

    Keyboard.dismiss();

    // reset state
    this.setState({
      date: this.getDate(),
    });

    if (Platform.OS === 'ios') {
      this.setModalVisible(true);
    } else {
      const {
        mode,
      } = this.props;
      if (mode === 'date' || mode === 'datetime') {
        this.setAndroidDatePickerVisible(true);
      } else if (mode === 'time') {
        this.setAndroidTimePickerVisible(true);
      }
    }

    if (typeof this.props.onOpenModal === 'function') {
      this.props.onOpenModal();
    }
  }

  _renderIcon() {
    const {
      showIcon,
      iconSource,
      iconComponent,
      customStyles
    } = this.props;

    if (showIcon) {
      if (iconComponent) {
        return iconComponent;
      }
      return (
        <Image
          style={[Style.dateIcon, customStyles.dateIcon]}
          source={iconSource}
        />
      );
    }

    return null;
  }

  render() {
    const {
      mode,
      style,
      customStyles,
      height,
      disabled,
      minDate,
      maxDate,
      minuteInterval,
      timeZoneOffsetInMinutes,
      cancelBtnText,
      confirmBtnText,
      TouchableComponent,
      testID,
      cancelBtnTestID,
      confirmBtnTestID,
      allowFontScaling,
      locale,
    } = this.props;

    const position= {
      bottom: -height,
      transform: [{ translateY: this.state.animatedPosition }],
    };

    const {
      androidDatePickerVisible,
      androidTimePickerVisible,
    } = this.state;

    const { androidMode, format = FORMATS[mode], is24Hour = !format.match(/h|a/) } = this.props;

    const dateInputStyle = [
      Style.dateInput, customStyles.dateInput,
      disabled && Style.disabled,
      disabled && customStyles.disabled,
    ];

    return (
      <TouchableComponent
        style={[Style.dateTouch, style]}
        underlayColor={'transparent'}
        onPress={this.onPressDate}
        testID={testID}
      >
        <View style={[Style.dateTouchBody, customStyles.dateTouchBody]}>
          {
            !this.props.hideText ?
              <View style={dateInputStyle}>
                {this.getTitleElement()}
              </View>
              :
              <View/>
          }
          {this._renderIcon()}
          {Platform.OS === 'ios' && <Modal
            transparent
            animationType="none"
            visible={this.state.modalVisible}
            supportedOrientations={SUPPORTED_ORIENTATIONS}
            onRequestClose={() => { this.setModalVisible(false); }}
          >
            <View
              style={{flex: 1}}
            >
              <TouchableComponent
                style={Style.datePickerMask}
                activeOpacity={1}
                underlayColor={'#00000077'}
                onPress={this.onPressMask}
              >
                <TouchableComponent
                  underlayColor={'#fff'}
                  style={{flex: 1}}
                >
                  <Animated.View
                    style={[Style.datePickerCon, position, { height }, customStyles.datePickerCon]}
                  >
                    <View pointerEvents={this.state.allowPointerEvents ? 'auto' : 'none'}>
                      <RNDateTimePicker
                        value={Moment(this.state.date).toDate()}
                        mode={mode}
                        minimumDate={minDate && this.getDate(minDate)}
                        maximumDate={maxDate && this.getDate(maxDate)}
                        onChange={this.onDateChange}
                        minuteInterval={minuteInterval}
                        timeZoneOffsetInMinutes={timeZoneOffsetInMinutes ? timeZoneOffsetInMinutes : null}
                        style={[Style.datePicker, customStyles.datePicker]}
                        locale={locale}
                      />
                    </View>
                    <TouchableComponent
                      underlayColor={'transparent'}
                      onPress={this.onPressCancel}
                      style={[Style.btnText, Style.btnCancel, customStyles.btnCancel]}
                      testID={cancelBtnTestID}
                    >
                      <Text
                        allowFontScaling={allowFontScaling}
                        style={[Style.btnTextText, Style.btnTextCancel, customStyles.btnTextCancel]}
                      >
                        {cancelBtnText}
                      </Text>
                    </TouchableComponent>
                    <TouchableComponent
                      underlayColor={'transparent'}
                      onPress={this.onPressConfirm}
                      style={[Style.btnText, Style.btnConfirm, customStyles.btnConfirm]}
                      testID={confirmBtnTestID}
                    >
                      <Text allowFontScaling={allowFontScaling}
                            style={[Style.btnTextText, customStyles.btnTextConfirm]}
                      >
                        {confirmBtnText}
                      </Text>
                    </TouchableComponent>
                  </Animated.View>
                </TouchableComponent>
              </TouchableComponent>
            </View>
          </Modal>}
          {(androidDatePickerVisible && Platform.OS === 'android' && mode === 'date') && <RNDateTimePicker
            mode="date"
            display={androidMode}
            is24Hour={is24Hour}
            minimumDate={minDate && this.getDate(minDate)}
            maximumDate={maxDate && this.getDate(maxDate)}
            value={Moment(this.state.date).toDate()}
            onChange={this.onDatePicked}
          />
          }

          {(androidTimePickerVisible && Platform.OS === 'android' && mode === 'time') &&
          <RNDateTimePicker
            mode="time"
            display={androidMode}
            is24Hour={is24Hour}
            value={Moment(this.state.date).toDate()}
            onChange={this.onTimePicked}
          />
          }

          {(androidDatePickerVisible && Platform.OS === 'android' && mode === 'datetime') &&
          <RNDateTimePicker
            mode="date"
            display={androidMode}
            minimumDate={minDate && this.getDate(minDate)}
            maximumDate={maxDate && this.getDate(maxDate)}
            value={Moment(this.state.date).toDate()}
            onChange={this.onDatetimePicked}
          />
          }
          {(androidTimePickerVisible &&  Platform.OS === 'android' && mode === 'datetime') &&
          <RNDateTimePicker
            mode="time"
            display={androidMode}
            is24Hour={is24Hour}
            value={Moment(this.state.date).toDate()}
            onChange={this.onDatetimeTimePicked}
          />
          }
        </View>
      </TouchableComponent>
    );
  }
}

DatePicker.defaultProps = {
  mode: 'date',
  androidMode: 'default',
  date: '',
  // component height: 216(DatePickerIOS) + 1(borderTop) + 42(marginTop), IOS only
  height: 259,

  // slide animation duration time, default to 300ms, IOS only
  duration: 300,
  confirmBtnText: 'Confifmm',
  cancelBtnText: 'Cancel',
  iconSource: require('./date_icon.png'),
  customStyles: {},

  // whether or not show the icon
  showIcon: true,
  disabled: false,
  allowFontScaling: true,
  hideText: false,
  placeholder: '',
  TouchableComponent: TouchableHighlight,
  modalOnResponderTerminationRequest: e => true
};

DatePicker.propTypes = {
  mode: PropTypes.oneOf(['date', 'datetime', 'time']),
  androidMode: PropTypes.oneOf(['clock', 'calendar', 'spinner', 'default']),
  date: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date), PropTypes.object]),
  format: PropTypes.string,
  minDate: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date)]),
  maxDate: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date)]),
  height: PropTypes.number,
  duration: PropTypes.number,
  confirmBtnText: PropTypes.string,
  cancelBtnText: PropTypes.string,
  iconSource: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
  iconComponent: PropTypes.element,
  customStyles: PropTypes.object,
  showIcon: PropTypes.bool,
  disabled: PropTypes.bool,
  allowFontScaling: PropTypes.bool,
  onDateChange: PropTypes.func,
  onOpenModal: PropTypes.func,
  onCloseModal: PropTypes.func,
  onPressMask: PropTypes.func,
  placeholder: PropTypes.string,
  modalOnResponderTerminationRequest: PropTypes.func,
  is24Hour: PropTypes.bool,
  getDateStr: PropTypes.func,
  locale: PropTypes.string
};

export default DatePicker;
